L41-node-server-async-bind.md

```diff
@@ -0,0 +1,46 @@
+Asynchronous port binding method in Node Server API
+----
+* Author(s): murgatroid99
+* Approver: wenbozhu
+* Status: Draft
+* Implemented in: Node
+* Last updated: 28-09-2018
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Uio1Ja6GRHI
+
+## Abstract
+
+Add a method to the Node gRPC `Server` class that binds a port as an asynchronous operation.
+
+## Background
+
+In the initial development of the Node gRPC library, the `Server` class method to bind a port, called `bind`, was implemented as a synchronous operation because it was available as an asynchronous operation in the underlying core library. In contrast, in the Node built in `net` module, the `Server` class's `listen` method, which performs a similar function, is asynchronous. In the new pure JavaScript Node gRPC library, the underlying network operations use the `net` module, so it is impossible to implement the `bind` method as a synchronous operation.
+
+
+## Proposal
+
+We would add a method called `bindAsync` to the `Server` class that implements the same functionality as `bind`, but provides the result asynchronously. The specific function signature would be as follows:
+
+```ts
+class Server {
+
+  // ...
+
+  /**
+   * @param port The port that the server should bind on, in the format "address:port"
+   * @param creds Server credentials object to be used for SSL.
+   * @param callback Callback to be called asynchronously with the result of binding the port
+   */
+  bindAsync(port: string, creds: ServerCredentials, callback: (port: number)=>void): void;
```

Comment:
I am on the fence about that. On one hand, you make a good point about `util.promisify()` and consistency with other Node operations. On the other hand, the current `bind` API communicates failure with specific return values instead of throwing errors, so passing an error to the callback would be a greater semantic deviation from the existing API.

Reply:
It's your call. But if your implementation doesn't end up being compatible with `util.promisify()` out of the box, you might want to look into [custom promisified functions](https://nodejs.org/api/util.html#util_custom_promisified_functions).

Reply:
Actually, after thinking about it more I think we can have it both ways, or close enough. If we pass an error to the callback on failure, and also pass the number unconditionally, then users who want to can do the same checks with the number that they do now, and we will have compatibility with `util.promisify()`.